Order sync from hook

thisiscab opened the pull request:
Since it would be favorable to not run a rake task at a specific interval each day in order to sync the Shopify order to the Solidus order, we would like to have it link to a hook event.

Shopify allow us to set URL to calls when specific events happen, in this case, after an order gets completed on Shopify, we are calling a specific route that will take care of importing that result down to Solidus.

thisiscab committed:
Namespace pos order generator

We want the retail specific code under the retail namespace.

thisiscab committed:
Add the shared secret attributes for the configuration object

We need that property in order to validate if the webhook request
actually comes from Shopify.

thisiscab committed:
Add the shopify webhook controller

This will be used for all the webhook controllers that depends on
Shopify. Less code duplication yay!

thisiscab committed:
Add the order controller for hooking Shopify

This is the basic scaffolding that will eventually create a Solidus
order when receiving a Shopify create order hook event.

thisiscab committed:
Ensure that the hooks order controller calls the generator

thisiscab commented:
@bryanmtl @braidn @DanielWright 
What do you think of that?

thisiscab committed:
Proper namespace the failing spec

This was forgotten when changing the namespace for that specific
object

thisiscab committed:
Add the webhook shared secret

There are two different concept of shared secret from Shopify. One of
the key comes from the app while the webhook one comes from the webhook
configuration page.

thisiscab committed:
Fix how we fetch the ID of the order from the hook

thisiscab committed:
Merge branch 'master' into feature/shopify/order-sync-from-hook

DanielWright commented on app/controllers/spree/retail/shopify/hooks_controller.rb:
```diff
@@ -0,0 +1,28 @@
+        end
+
+        def json_body
+          @_json_body ||= JSON.parse(request.body.read)
```
Why doesn't Rails automagically parse this into `params`?

DanielWright replied:
I think you can fix this with `wrap_parameters :payload, format: :json`.

DanielWright replied:
I believe this happens because Rails assumes POST requests are form-data, not JSON. You might also be able to address this by specifying a default format of JSON when drawing the route.

DanielWright commented on app/controllers/spree/retail/shopify/hooks_controller.rb:
```diff
@@ -0,0 +1,28 @@
+          head :unauthorized unless request_authentified?(data, Spree::Retail::Config.shopify_webhook_shared_secret)
+        end
+
+        def request_authentified?(data, secret)
```
What do you think of encapsulating this logic into a nice `ShopifyPayload` object with an elegant interface (e.g. `#valid?`, etc.)?

Additionally, I think `authenticated` is the word you're looking for here. 😄

braidn replied:
I agree with Dan here, 'authentified?' is not your friend.

DanielWright reviewed (changes requested):
Having to read the body from the `request` object is the main smell to me here.

braidn commented on spec/support/shared/shopify_request.rb:
```diff
@@ -15,13 +15,17 @@
-    file = endpoint.split('/')[0]
     url = "#{ShopifyAPI::Base.site}/#{endpoint}.#{extension}"
-    json = File.open("#{File.dirname(__FILE__)}/../data/#{file}.#{extension}")
+    json = read_file_from_endpoint(endpoint, extension)
```
This is not really what's happening. This also means this abstractions name doesn't clue people into how to reuse this helper method.

braidn reviewed (commented):
💥 Charles. Code 👁 ed

thisiscab committed:
Try to allow shopify hook request

It seems like Solidus is blocking the request because it has not been
authentified. This should solve the problem.

thisiscab committed:
Merge branch 'master' into feature/shopify/order-sync-from-hook

thisiscab committed:
Re-enable Shopify API authentication

Else anyone could create fake orders which would seriously causes major
problems

thisiscab committed:
Add a service class that verifies the ShopifyPayload

thisiscab committed:
Use the params to fetch the requests params

thisiscab commented:
Is that better guys @DanielWright @braidn

thisiscab committed:
Merge branch 'master' into feature/shopify/order-sync-from-hook

thisiscab committed:
Merge branch 'master' into feature/shopify/order-sync-from-hook
